docs/design-docs/03-pipeline-merge.md

```diff
@@ -0,0 +1,405 @@
+        tag: 1.0.1
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.0.1
+        new-setter: new-setter-value // new setter is added
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.2.0 // value of tag is updated
+
+Current Output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.0.1 // entire pipeline is overridden by upstream 
+        new-setter: new-setter-value
+
+Expected Output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configMap:
+        image: nginx
+        tag: 1.2.0 // updated tag is preserved
+        new-setter: new-setter-value // new setter is pulled
+```
+
+In the above scenario, the value of the setter tag is not updated upstream, so 
+the modified local value will be preserved. But in the following example, both 
+upstream and local values change. So, similar to merging resources, upstream 
+value wins if the same fields in both upstream and local are updated.
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters-updated.yaml // upstream value changed
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters-local.yaml // local value changed
+
+Expected Output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters-updated.yaml // upstream overrides local 
+```
+
+Similarly, the upstream version wins if both upstream and local are updated, else local is preserved.
+
+```yaml
+Original:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1
+      configPath: setters.yaml
+
+Updated upstream:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1.2
+      configPath: setters.yaml
+
+Local:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1.1
+      configPath: setters.yaml
+
+Expected Output:
+
+pipeline:
+  mutators:
+    - image: gcr.io/kpt-fn/apply-setters:v0.1.2
+      configPath: setters-updated.yaml
+```
+
+This might not be what all users expect. But this is the default behavior in case 
+of conflict while merging normal resources as well. So we can inherit the behavior 
+and be consistent. In order to provide more visibility to the users, we can add 
+log messages in cases of such conflicts and intimate users about the updated value. 
+In the future, we can add support to a different conflict strategy of `--local-wins`
+as an option to the kpt pkg update command.
+
+#### More examples with expected output
+
+Newly added upstream functions are appended at the end.
```

Comment:
The biggest challenge about merging lists, is handling addition, deleting and moved elements. I don't think we'll be able to come up with rules here that will always be what users want, but we need to make sure we have clear rules that understandable and well documented. I think we need to express this more concise than only with examples.

Reply:
Yes. We defined the merge keys in the order of precedence. We will use default merge behavior.

Reply:
I still think we need a clearer explanation of the behavior here. It is not obvious how we handle scenarios where only some functions have the name specified and how the relative ordering actually impacts the resulting merge.